torchvision/csrc/ops/quantized/cpu/qnms_kernel.cpp

```diff
@@ -0,0 +1,126 @@
+#include <ATen/ATen.h>
+#include <ATen/native/quantized/affine_quantizer.h>
+#include <torch/library.h>
+
+namespace vision {
+namespace ops {
+
+namespace {
+
+template <typename scalar_t>
+at::Tensor qnms_kernel_impl(
+    const at::Tensor& dets,
+    const at::Tensor& scores,
+    double iou_threshold) {
+
+  TORCH_CHECK(!dets.is_cuda(), "dets must be a CPU tensor");
+  TORCH_CHECK(!scores.is_cuda(), "scores must be a CPU tensor");
+  TORCH_CHECK(
+      dets.scalar_type() == scores.scalar_type(),
+      "dets should have the same type as scores");
+
+  if (dets.numel() == 0)
+    return at::empty({0}, dets.options().dtype(at::kLong)); 
+
+  const auto ndets = dets.size(0);
+
+  auto x1_t = dets.select(1, 0).contiguous();
+  auto y1_t = dets.select(1, 1).contiguous();
+  auto x2_t = dets.select(1, 2).contiguous();
+  auto y2_t = dets.select(1, 3).contiguous();
+  auto order_t = std::get<1>(scores.sort(0, /* descending=*/true));
+  at::Tensor suppressed_t = at::zeros({ndets}, dets.options().dtype(at::kByte));
+  at::Tensor keep_t = at::zeros({ndets}, dets.options().dtype(at::kLong));
+  at::Tensor areas_t = at::zeros({ndets}, dets.options().dtype(at::kFloat));
+
+  auto suppressed = suppressed_t.data_ptr<uint8_t>();
+  auto keep = keep_t.data_ptr<int64_t>();
+  auto order = order_t.data_ptr<int64_t>();
+  auto x1 = x1_t.data_ptr<scalar_t>();
+  auto y1 = y1_t.data_ptr<scalar_t>();
+  auto x2 = x2_t.data_ptr<scalar_t>();
+  auto y2 = y2_t.data_ptr<scalar_t>();
+  auto areas = areas_t.data_ptr<float>();
+
+  for (int64_t i = 0; i < ndets; i++) {
+    // Note 1: To get the exact area we'd need to multiply by scale**2, but this
+    // would get canceled out in the computation of ovr below. So we leave that
+    // out.
+    // Note 2: degenerate boxes (x2 < x1 or y2 < y1) may underflow. Same below
+    // when computing w and h
+    areas[i] = (x2[i].val_ - x1[i].val_) * (y2[i].val_ - y1[i].val_);
+  }
+
+  int64_t num_to_keep = 0;
+
+  for (int64_t _i = 0; _i < ndets; _i++) {
+    auto i = order[_i];
+    if (suppressed[i] == 1)
+      continue;
+    keep[num_to_keep++] = i;
+
+    auto ix1val = x1[i].val_;
+    auto iy1val = y1[i].val_;
+    auto ix2val = x2[i].val_;
+    auto iy2val = y2[i].val_;
+    auto iarea = areas[i];
+
+    for (int64_t _j = _i + 1; _j < ndets; _j++) {
+      auto j = order[_j];
+      if (suppressed[j] == 1)
+        continue;
+      auto xx1 = std::max(ix1val, x1[j].val_);
+      auto yy1 = std::max(iy1val, y1[j].val_);
+      auto xx2 = std::min(ix2val, x2[j].val_);
+      auto yy2 = std::min(iy2val, y2[j].val_);
+
+      auto w = std::max(0, xx2 - xx1);  // * scale (gets canceled below)
+      auto h = std::max(0, yy2 - yy1);  // * scale (gets canceled below)
```

Comment:
In fact I think it's possible to get an underflow on unsigned types here, even when boxes aren't degenerate: if box j is strictly left from box i and they don't overlap, then we'll have `xx2 < xx1`.

Reply:
yes, good catch, this happens more often than just with degenerate boxes here

Reply:
Trying to come up with a reproducible example of an underflow was quite informative, and I came to the conclusion that:

1) they actually don't happen
2) Even when they do (by manually forcing them to happen), they're effectively harmless, i.e. they don't change the end result. Since they don't change the result, it's actually impossible to test against underflow.

The reason for 1) is that the result of `a - b` is an `int`, even if both `a` and `b` are `uint8_t`: TIL about integral promotion (https://stackoverflow.com/questions/21421998/subtraction-operation-on-unsigned-char). Since `quint_8` is the only supported quantized unsigned type, such values can always be represented as `int`s, and integral promotion will always happen.

Regarding 2): It's possible to get an underflow by doing some explicit casts to `uint8_t` in the `qnms` code. The following boxes and scores:

```py
    boxes = torch.as_tensor([[0, 0, 5, 5], [10, 0, 20, 20]]).to(torch.float)
    scores = torch.as_tensor([10, 9]).to(torch.float)
```

should produce `10 5 0 5` for `xx1, xx2, yy1, yy2` and thus lead to an underflow when computing `w = xx2 - xx1`.

However, the underflow leads to `w = 251` and `inter = some_big_value = 1255` and thus `ovr ~= -2` (note the negative sign because of `some_big_value` that gets subtracted to the denominator).

Hence the `if (ovr > iou_threshold)` condition is never verified when there's an underflow, and that's actually correct since the correct value for `ovr` should have been `0` anyway: underflows can only happen when the boxes are disjoint along at least one of the dimensions.

So the results are unchanged, and it's thus impossible to actually test against underflows... I think.